crypto/ec/ec_asn1.c

```diff
@@ -485,7 +485,7 @@ ECPKPARAMETERS *EC_GROUP_get_ecpkparameters(const EC_GROUP *group,
             ECPARAMETERS_free(ret->value.parameters);
     }
 
-    if (EC_GROUP_get_asn1_flag(group)) {
+    if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE) {
```

Comment:
Shouldn't this be `&` not `==` here and throughout?

Existing code uses both forms: `==` once and `&` twice.

Reply:
Uhm, this ASN1 flag is actually a single bit, so there shouldn't be a difference. I can go either way. In 1.1.1 this flag is never associated with boolean operators.

Reply:
I actually prefer `==`, it allows symmetry between

~~~C
if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE)
~~~

and 

~~~C
if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_EXPLICIT_CURVE)
~~~

We cannot use boolean operators on `OPENSSL_EC_EXPLICIT_CURVE` as its value is `0x0`

Reply:
Also, we spread around code like

~~~C
EC_GROUP_set_asn1_flag(ret_group, OPENSSL_EC_EXPLICIT_CURVE)
~~~

so I believe there is little point in overengineering for a future where this flag field will ever include additional bits: changing it in the future will break a lot of code.

Reply:
This patch would improve consistency. I dislike the first hunk.

~~~diff
diff --git a/crypto/ec/ec_backend.c b/crypto/ec/ec_backend.c
index 381da71f33..4ee093f06b 100644
--- a/crypto/ec/ec_backend.c
+++ b/crypto/ec/ec_backend.c
@@ -308,7 +308,8 @@ int ossl_ec_group_todata(const EC_GROUP *group, OSSL_PARAM_BLD *tmpl,
         ERR_raise(ERR_LIB_EC, EC_R_INVALID_FORM);
         return 0;
     }
-    encoding_flag = EC_GROUP_get_asn1_flag(group) & OPENSSL_EC_NAMED_CURVE;
+    encoding_flag = EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE ?
+        OPENSSL_EC_NAMED_CURVE : OPENSSL_EC_EXPLICIT_CURVE;
     encoding_name = ec_param_encoding_id2name(encoding_flag);
     if (encoding_name == NULL
         || !ossl_param_build_set_utf8_string(tmpl, params,
diff --git a/providers/implementations/encode_decode/encode_key2any.c b/providers/implementations/encode_decode/encode_key2any.c
index 855866cbfe..1ade5800bb 100644
--- a/providers/implementations/encode_decode/encode_key2any.c
+++ b/providers/implementations/encode_decode/encode_key2any.c
@@ -679,7 +679,7 @@ static int prepare_ec_params(const void *eckey, int nid, int save,
     }
 
     if (curve_nid != NID_undef
-        && (EC_GROUP_get_asn1_flag(group) & OPENSSL_EC_NAMED_CURVE)) {
+        && (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE)) {
         /* The CHOICE came to namedCurve */
         if (OBJ_length(params) == 0) {
             /* Some curves might not have an associated OID */
diff --git a/providers/implementations/encode_decode/encode_key2text.c b/providers/implementations/encode_decode/encode_key2text.c
index f8f9712e96..60f3e13a40 100644
--- a/providers/implementations/encode_decode/encode_key2text.c
+++ b/providers/implementations/encode_decode/encode_key2text.c
@@ -460,7 +460,7 @@ err:
 static int ec_param_to_text(BIO *out, const EC_GROUP *group,
                             OSSL_LIB_CTX *libctx)
 {
-    if (EC_GROUP_get_asn1_flag(group) & OPENSSL_EC_NAMED_CURVE) {
+    if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE) {
         const char *curve_name;
         int curve_nid = EC_GROUP_get_curve_name(group);
 
~~~

Reply:
If we're just using it for this one flag, the function/field names are wrong.  _named_ would make more sense.

Reply:
I agree the name is not particularly friendly, especially given that other similar functions are named as `*_get_flags()` and handled with bitwise logic. "asn1_flag" is signular, but indeed not very descriptive.

It seems unfortunately it is too late to change this now. Looking back at our history, it seems quite a deliberate design the fact that this is a 1-bit flag.

Before 254ef80db19a6f6610ef07535efd699cef7aa855 it had more ~~bits~~ values —although admittedly a bit messy as it was somehow possible to be both `OPENSSL_EC_EXPLICIT` and `OPENSSL_EC_NAMED_CURVE`— and was used also to pick the default encodings for points among uncompressed, compressed and hybrid forms.
Before that we were using bitwise logic to handle the flag field, masking out bits of it in groups.
At this point it became a 1-bit field, and later on with 86f300d38540ead85543aee0cb30c32145931744, we started replacing `0` with a more descriptive `OPENSSL_EC_EXPLICIT_CURVE`.
Code expecting that field to be exactly 1-bit is now being out for so long that I am not sure we can do much about it.

Ideally we could have had this discussion much sooner, a little bit over 1 year ago when the issue was first opened, so we could have included `EC_GROUP_get_asn1_flag()` and `EC_GROUP_set_asn1_flag()` in the deprecation list for 3.0.0, while providing a new better named function as an alternative.
I don't think this is something we can do this close to beta, unfortunately, so we will likely have to stuck with it until the next major release :(

Reply:
@paulidale how should I address the comment? do you still prefer bitwise access? If instead we go for `==` I'd rather do the consistency improvement of https://github.com/openssl/openssl/pull/16355#discussion_r691689240 as a separate PR.

Reply:
Go for `==` and add the consistency separately.